File: backend/src/routes/relationships.ts
import { Router, Request, Response } from 'express';
import Relationship from '../models/Relationship';
import User from '../models/User';
import { authMiddleware, roleMiddleware } from '../middleware/auth';

const router = Router();

// GET /api/relationships – Get hierarchical relationships: caregivers -> patients -> family members
router.get('/', authMiddleware, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.userId;
    const role = req.user!.role;
    
    let relationships: any[] = [];

    if (role === 'super_admin') {
      // Super admin sees all caregivers across all corporates with their patients and families
      const caregivers = await User.find({ role: 'caregiver' }).select('_id full_name email corporate_id');
      
      for (const caregiver of caregivers) {
        const patients = await User.find({ caregiver_id: caregiver._id }).select('_id full_name email family_ids');
        const patientsList = [];
        
        for (const patient of patients) {
          const familyMembers = patient.family_ids && patient.family_ids.length > 0
            ? await User.find({ _id: { $in: patient.family_ids } }).select('_id full_name email role')
            : [];
          
          patientsList.push({
            patient: { _id: patient._id, full_name: patient.full_name, email: patient.email },
            family_members: familyMembers.map((f: any) => ({ _id: f._id, full_name: f.full_name, email: f.email, role: f.role }))
          });
        }
        
        relationships.push({
          caregiver: { _id: caregiver._id, full_name: caregiver.full_name, email: caregiver.email, corporate_id: caregiver.corporate_id },
          patients: patientsList
        });
      }
    } else if (role === 'admin') {
      // Admin sees caregivers under their corporate_id with their patients and families
      const caregivers = await User.find({ role: 'caregiver', corporate_id: userId }).select('_id full_name email');
      
      for (const caregiver of caregivers) {
        const patients = await User.find({ caregiver_id: caregiver._id }).select('_id full_name email family_ids');
        const patientsList = [];
        
        for (const patient of patients) {
          const familyMembers = patient.family_ids && patient.family_ids.length > 0
            ? await User.find({ _id: { $in: patient.family_ids } }).select('_id full_name email role')
            : [];
          
          patientsList.push({
            patient: { _id: patient._id, full_name: patient.full_name, email: patient.email },
            family_members: familyMembers.map((f: any) => ({ _id: f._id, full_name: f.full_name, email: f.email, role: f.role }))
          });
        }
        
        relationships.push({
          caregiver: { _id: caregiver._id, full_name: caregiver.full_name, email: caregiver.email },
          patients: patientsList
        });
      }
    } else if (role === 'caregiver') {
      // Caregiver sees their own patients and families
      const patients = await User.find({ caregiver_id: userId }).select('_id full_name email family_ids');
      const patientsList = [];
      
      for (const patient of patients) {
        const familyMembers = patient.family_ids && patient.family_ids.length > 0
          ? await User.find({ _id: { $in: patient.family_ids } }).select('_id full_name email role')
          : [];
        
        patientsList.push({
          patient: { _id: patient._id, full_name: patient.full_name, email: patient.email },
          family_members: familyMembers.map((f: any) => ({ _id: f._id, full_name: f.full_name, email: f.email, role: f.role }))
        });
      }
      
      const currentCaregiver = await User.findById(userId).select('_id full_name email');
      relationships.push({
        caregiver: { _id: currentCaregiver?._id, full_name: currentCaregiver?.full_name, email: currentCaregiver?.email },
        patients: patientsList
      });
    } else if (role === 'patient') {
      // Patient sees their own caregiver and family members
      const currentPatient = await User.findById(userId).populate('caregiver_id', '_id full_name email').populate('family_ids', '_id full_name email role');
      
      if (currentPatient?.caregiver_id) {
        relationships.push({
          caregiver: currentPatient.caregiver_id,
          patients: [
            {
              patient: { _id: currentPatient._id, full_name: currentPatient.full_name, email: currentPatient.email },
              family_members: Array.isArray(currentPatient.family_ids) ? currentPatient.family_ids : []
            }
          ]
        });
      }
    }

    res.json(relationships);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/relationships
router.post('/', authMiddleware, roleMiddleware('admin', 'caregiver'), async (req: Request, res: Response) => {
  try {
    const { patient_id, related_user_id, relationship_type } = req.body;

    // Check for duplicate
    const existing = await Relationship.findOne({ patient_id, related_user_id });
    if (existing) {
      res.status(400).json({ error: 'Relationship already exists' });
      return;
    }

    const relationship = await Relationship.create({
      patient_id,
      related_user_id,
      relationship_type,
      created_by: req.user!.userId,
    });

    res.status(201).json(relationship);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// DELETE /api/relationships/:id
router.delete('/:id', authMiddleware, roleMiddleware('admin'), async (req: Request, res: Response) => {
  try {
    await Relationship.findByIdAndDelete(req.params.id);
    res.json({ success: true });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

export default router;

File: backend-py/app/routes/relationships.py
import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.relationship import Relationship
from app.models.user import User
from app.auth import get_current_user, RoleChecker

router = APIRouter(prefix="/api/relationships", tags=["relationships"])


# GET /api/relationships
@router.get("/")
def list_relationships(
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    role = current_user["role"]
    user_id = current_user["userId"]
    relationships = []

    if role == "super_admin":
        caregivers = db.query(User).filter(User.role == "caregiver").all()
        for cg in caregivers:
            patients = db.query(User).filter(User.caregiver_id == cg.id).all()
            patients_list = []
            for p in patients:
                fam = []
                if p.family_ids:
                    fam = db.query(User).filter(User.id.in_(p.family_ids)).all()
                patients_list.append({
                    "patient": {"_id": str(p.id), "full_name": p.full_name, "email": p.email},
                    "family_members": [{"_id": str(f.id), "full_name": f.full_name, "email": f.email, "role": f.role} for f in fam],
                })
            relationships.append({
                "caregiver": {"_id": str(cg.id), "full_name": cg.full_name, "email": cg.email, "corporate_id": str(cg.corporate_id) if cg.corporate_id else None},
                "patients": patients_list,
            })

    elif role == "admin":
        caregivers = db.query(User).filter(User.role == "caregiver", User.corporate_id == uuid.UUID(user_id)).all()
        for cg in caregivers:
            patients = db.query(User).filter(User.caregiver_id == cg.id).all()
            patients_list = []
            for p in patients:
                fam = db.query(User).filter(User.id.in_(p.family_ids)).all() if p.family_ids else []
                patients_list.append({
                    "patient": {"_id": str(p.id), "full_name": p.full_name, "email": p.email},
                    "family_members": [{"_id": str(f.id), "full_name": f.full_name, "email": f.email, "role": f.role} for f in fam],
                })
            relationships.append({
                "caregiver": {"_id": str(cg.id), "full_name": cg.full_name, "email": cg.email},
                "patients": patients_list,
            })

    elif role == "caregiver":
        cg = db.query(User).filter(User.id == uuid.UUID(user_id)).first()
        patients = db.query(User).filter(User.caregiver_id == uuid.UUID(user_id)).all()
        patients_list = []
        for p in patients:
            fam = db.query(User).filter(User.id.in_(p.family_ids)).all() if p.family_ids else []
            patients_list.append({
                "patient": {"_id": str(p.id), "full_name": p.full_name, "email": p.email},
                "family_members": [{"_id": str(f.id), "full_name": f.full_name, "email": f.email, "role": f.role} for f in fam],
            })
        relationships.append({
            "caregiver": {"_id": str(cg.id), "full_name": cg.full_name, "email": cg.email} if cg else {},
            "patients": patients_list,
        })

    elif role == "patient":
        pt = db.query(User).filter(User.id == uuid.UUID(user_id)).first()
        if pt and pt.caregiver_id:
            cg = db.query(User).filter(User.id == pt.caregiver_id).first()
            fam = db.query(User).filter(User.id.in_(pt.family_ids)).all() if pt.family_ids else []
            relationships.append({
                "caregiver": {"_id": str(cg.id), "full_name": cg.full_name, "email": cg.email} if cg else {},
                "patients": [{
                    "patient": {"_id": str(pt.id), "full_name": pt.full_name, "email": pt.email},
                    "family_members": [{"_id": str(f.id), "full_name": f.full_name, "email": f.email, "role": f.role} for f in fam],
                }],
            })

    elif role == "family":
        from app.models.relationship import Relationship as RelModel
        fam_rels = (
            db.query(RelModel)
            .filter(
                RelModel.related_user_id == uuid.UUID(user_id),
                RelModel.relationship_type == "family",
            )
            .all()
        )
        for rel in fam_rels:
            pt = db.query(User).filter(User.id == rel.patient_id).first()
            if not pt:
                continue
            cg = db.query(User).filter(User.id == pt.caregiver_id).first() if pt.caregiver_id else None
            relationships.append({
                "caregiver": {"_id": str(cg.id), "full_name": cg.full_name, "email": cg.email} if cg else {},
                "patients": [{
                    "patient": {"_id": str(pt.id), "full_name": pt.full_name, "email": pt.email},
                    "family_members": [],
                }],
            })

    return relationships


# POST /api/relationships
@router.post("/", status_code=201)
def create_relationship(
    body: dict,
    current_user: dict = Depends(RoleChecker(["admin", "caregiver"])),
    db: Session = Depends(get_db),
):
    patient_id = uuid.UUID(body["patient_id"])
    related_user_id = uuid.UUID(body["related_user_id"])

    existing = db.query(Relationship).filter(
        Relationship.patient_id == patient_id,
        Relationship.related_user_id == related_user_id,
    ).first()
    if existing:
        raise HTTPException(400, "Relationship already exists")

    rel = Relationship(
        patient_id=patient_id,
        related_user_id=related_user_id,
        relationship_type=body["relationship_type"],
        created_by=uuid.UUID(current_user["userId"]),
    )
    db.add(rel)
    db.commit()
    db.refresh(rel)
    return rel.to_dict()


# DELETE /api/relationships/:id
@router.delete("/{rel_id}")
def delete_relationship(
    rel_id: str,
    current_user: dict = Depends(RoleChecker(["admin"])),
    db: Session = Depends(get_db),
):
    rel = db.query(Relationship).filter(Relationship.id == uuid.UUID(rel_id)).first()
    if not rel:
        raise HTTPException(404, "Relationship not found")
    db.delete(rel)
    db.commit()
    return {"success": True}
